pallets/corporate-actions/src/lib.rs

```diff
@@ -213,28 +320,133 @@ decl_module! {
+        /// - `details` of the CA in free-text form, up to a certain number of bytes in length.
+        /// - `targets`, if any, which this CA is relevant/irrelevant to.
+        ///    Overrides, if provided, the default at the asset level (`set_default_targets`).
+        /// - `default_wt`, if any, is the default withholding tax to use for this CA.
+        ///    Overrides, if provided, the default at the asset level (`set_default_withholding_tax`).
+        /// - `wt`, if any, provides per-DID withholding tax overrides.
+        ///    Overrides, if provided, the default at the asset level (`set_did_withholding_tax`).
+        ///
+        /// # Errors
+        /// - `DetailsTooLong` if `details.len()` goes beyond `max_details_length`.
+        /// - `UnauthorizedAsAgent` if `origin` is not `ticker`'s sole CAA (owner is not necessarily the CAA).
+        /// - `LocalCAIdOverflow` in the unlikely event that so many CAs were created for this `ticker`,
+        ///   that integer overflow would have occured if instead allowed.
+        #[weight = <T as Trait>::WeightInfo::initiate_corporate_action()]
+        pub fn initiate_corporate_action(
+            origin,
+            ticker: Ticker,
+            kind: CAKind,
+            record_date: Option<Moment>,
+            details: CADetails,
+            targets: Option<TargetIdentities>,
+            default_wt: Option<Tax>,
+            wt: Option<Vec<(IdentityId, Tax)>>,
+        ) {
+            // Ensure that `details` is short enough.
+            details
+                .len()
+                .try_into()
+                .ok()
+                .filter(|&len: &u32| len <= Self::max_details_length())
+                .ok_or(Error::<T>::DetailsTooLong)?;
+
+            // Ensure that CAA is calling.
+            let caa = Self::ensure_ca_agent(origin, ticker)?;
+
+            // Ensure that the next local CA ID doesn't overflow.
+            let local_id = CAIdSequence::get(ticker);
+            let next_id = local_id.0.checked_add(1).map(LocalCAId).ok_or(Error::<T>::LocalCAIdOverflow)?;
+            let id = CAId { ticker, local_id };
+
+            // Create a checkpoint at `record_date`, if any.
+            if let Some(record_date) = record_date {
+                <Asset<T>>::_create_checkpoint_emit(ticker, record_date, caa)?;
+            }
+
+            // Commit the next local CA ID.
+            CAIdSequence::insert(ticker, next_id);
+
+            // Use asset level defaults if data not provided here.
+            let targets = targets.unwrap_or_else(|| Self::default_target_identities(ticker));
+            let dwt = default_wt.unwrap_or_else(|| Self::default_withholding_tax(ticker));
+            let wt = match wt {
+                None => Self::did_withholding_tax(ticker),
+                Some(mut wt) => {
+                    // Nuke duplicates.
+                    wt.sort_unstable_by_key(|&(did, _)| did);
+                    wt.dedup_by_key(|&mut (did, _)| did);
```

Comment:
If someone enters `[(Alice, 10), (Alice, 20)]`, maybe we should return an error instead of deduplicating arbitrarily.